receiver/snmpreceiver/scraper.go

```diff
@@ -39,13 +90,489 @@ func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCrea
+	// basic ResourceMetrics. The info in this map is ultimately already
+	// contained within the resourceMetricsMap, but it is more easily accessible
+	// to pull out a specific existing Metric by resource and metric name using this
+	metricsByResource := newMetricsByResourceContainer()
+	metricsByResource.putMetrics(generalResourceKey, metricsMap)
+
+	// Try to scrape column OID based metrics.
+	// resourceMetricsMap is passed in as a place for any created resources/metrics to live
+	// metricsByResource is provided as an easy way to check if a metric is already
+	// associated with an existing resource
+	s.scrapeIndexedMetrics(now, resourceMetricsMap, metricsByResource, &scraperErrors)
+
+	// Return error if we failed to scrape any metrics
 	md := pmetric.NewMetrics()
 
-	return md, nil
+	// Put all of the created ResourceMetrics onto a top level Metrics
+	for _, value := range resourceMetricsMap {
+		value.MoveTo(md.ResourceMetrics().AppendEmpty())
+	}
+
+	return md, scraperErrors.Combine()
+}
+
+// scrapeScalarMetrics retrieves all SNMP data from scalar OIDs and turns the returned scalar data
+// into metrics with optional enum attributes
+func (s *snmpScraper) scrapeScalarMetrics(now pcommon.Timestamp, metricsMap map[string]*pmetric.Metric, scraperErrors *scrapererror.ScrapeErrors) {
+	scalarMetricNamesByOID := map[string]string{}
+	scalarMetricOIDs := []string{}
+
+	// Find all metric scalar OIDs
+	// Also create a map of metric names with OID as key so the metric config will be easy to
+	// matchup later with returned data
+	for name, metricCfg := range s.cfg.Metrics {
+		if len(metricCfg.ScalarOIDs) > 0 {
+			for i, oid := range metricCfg.ScalarOIDs {
+				// Data is returned by the client with '.' prefix on the OIDs.
+				// Making sure the prefix exists here in the configs so we can match it up with returned data later
+				if !strings.HasPrefix(oid.OID, ".") {
+					oid.OID = "." + oid.OID
+					s.cfg.Metrics[name].ScalarOIDs[i].OID = oid.OID
+				}
+				scalarMetricOIDs = append(scalarMetricOIDs, oid.OID)
+				scalarMetricNamesByOID[oid.OID] = name
+			}
+		}
+	}
+
+	// If no scalar metric configs, nothing else to do
+	if len(scalarMetricOIDs) == 0 {
+		return
+	}
+
+	// Get all SNMP scalar OID data and turn it into metrics/attributes
+	// which are then stored in the passed in metricsMap
+	s.client.GetScalarData(scalarMetricOIDs, scalarDataToMetric(now, metricsMap, scalarMetricNamesByOID, s.cfg), scraperErrors)
+}
+
+// scalarDataToMetric provides a function which will convert one piece of SNMP scalar data, turn it into
+// a metric with attributes based on the related configs, store it in the passed in metricsMap
+func scalarDataToMetric(
+	now pcommon.Timestamp,
+	metricsMap map[string]*pmetric.Metric,
+	scalarMetricNamesByOID map[string]string,
+	cfg *Config,
+) processFunc {
+	// This returns a processFunc because this is what the client's GetScalarData method requires
```

Comment:
@djaglowski i agree that this may not have been the best. it started with me looking at the gosnmp library and seeing how the primary walk function took an OID and a function to execute on the retrieved data. in my head at first this made sense from some sort of efficiency standpoint. by the time i didn't love this strategy...everything was already in place and it worked well....so i chose not to refactor.

Reply:
Is there a reason not to refactor then?

Reply:
@djaglowski i thought this response explained why i didn't refactor. 

to sum it up in the roughest of terms, following the pattern offered by the 3rd party internal client library was "good enough" at the end of the day.

if you disagree and think this design is unusable, let me know sooner than later as there will be a fair amount of work to redo.

Reply:
Yeah, this is a readability/maintainability problem to me. Ultimately, we have to establish a line between what we should do within the processing callback vs outside of it. I suspect the intention of that function is to apply _necessary_ processing that allows the walk to continue or stops the walk when possible.

As is, we're injecting most but not all processing into the callbacks, with the result being that it's difficult to decompose these functions. 

It sounds like there's not a technical reason to keep it this way, so we should refactor. I don't think it'll actually be that bad to shift the line so that the client returns `snmpData ` and then the scraper passes it to what we currently have as callbacks. From there, we'll be much freer to decompose the functions and reduce the amount of state that has to be passed around.